framework/src/meshgenerators/BreakMeshByBlockGenerator.C

```diff
@@ -114,6 +148,37 @@ BreakMeshByBlockGenerator::generate()
             connected_blocks.find(Elem::invalid_subdomain_id) != connected_blocks.end()
                 ? Elem::invalid_subdomain_id
                 : *subdomain_it;
+
+        // For the block_pairs option, the node that is only shared by specific block pairs will be
+        // newly created.
+        bool should_create_new_node = true;
+        if (_block_pairs_restricted)
+        {
+          auto elems = node_to_elem_map[current_node->id()];
+          for (auto elem_id = elems.begin(); elem_id != elems.end(); elem_id++)
+          {
+            subdomain_id_type block_id =
+                blockRestrictedElementSubdomainID(mesh->elem_ptr(*elem_id));
+            if (blockRestrictedElementSubdomainID(mesh->elem_ptr(*elem_id)) ==
+                Elem::invalid_subdomain_id)
+              should_create_new_node = false;
+
+            for (auto neighbor_elem_id = elem_id + 1; neighbor_elem_id != elems.end();
+                 neighbor_elem_id++)
+            {
+              subdomain_id_type neighbor_block_id =
+                  blockRestrictedElementSubdomainID(mesh->elem_ptr(*neighbor_elem_id));
+              if (block_id != neighbor_block_id && !findBlockPairs(block_id, neighbor_block_id))
+              {
+                should_create_new_node = false;
+                break;
+              }
+              if (!should_create_new_node)
+                break;
+            }
+          }
+        }
+
```

Comment:
I think this can be greatly simplified
the condition for which we decide to add a node is iff the domains attached to current_node are two, and if the two domains attached to current_node are in the block_pair_list then duplicate the node.
This means we can use an ordered set to capture the nodes and check if those two subdomains form a pair that is in the list.
Something like this should do the trick
```
if (_block_pairs_restricted){
   should_create_new_node = false
   std::set<subdomain_id_type> sets_blocks_for_this_node;
   for (auto elem_id = elems.begin(); elem_id != elems.end(); elem_id++)
      sets_blocks_for_this_node.insert(blockRestrictedElementSubdomainID(mesh->elem_ptr(*elem_id))
   if (sets_blocks_for_this_node == 2){
   {
      auto setIt = setOfStr.begin(); 
      if (findBlockPairs(setIt, std::next(setIt.begin(), 1)));
         should_create_new_node = true
   }
}
```

Reply:
seems like your logic works better. 

I made the suggested change.